torchci/docs/architecture.md

```diff
@@ -0,0 +1,82 @@
+# `torchci` architecture notes
+## Primary write path
+How does data get from GitHub to `torchci`?
+
+![torchci write path](write-path.png)
+
+Whenever something happens on GitHub, a [webhook event] is created and sent to
+all subscribers. The [webhook payload] will contain all the necessary
+information about the event. For example, if a new issue is created, an
+[`issues`] webhook is generated, providing information about the issue title,
+who created the issue, etc.
+
+[GitHub Apps] can subscribe to webhooks for the repos that they're installed
+on. `torchci` uses the [PyTorch Bot] app to keep track of what's going on in
+`pytorch/pytorch` and any other repos it's installed on.
+
+These webhooks are delivered to an API endpoint on [hud.pytorch.org]
+([`/api/github/webhooks`]), which writes the webhook payload to a DynamoDB
+table corresponding to the event type. For example, `workflow_job` payloads are
+written to `torchci-workflow-job`.
+
+Rockset is [integrated][dynamo] with DynamoDB and automatically picks up any
+changes in the tables in its collections.
+
+[webhook event]: https://docs.github.com/en/developers/webhooks-and-events/webhooks/about-webhooks
+[webhook payload]: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads
+[`issues`]: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#issues
+[GitHub Apps]: https://docs.github.com/en/developers/apps/getting-started-with-apps/about-apps
+[PyTorch Bot]: https://github.com/apps/pytorch-bot
+[`/api/github/webhooks`]: https://github.com/pytorch/test-infra/blob/main/torchci/pages/api/github/webhooks.ts
+[dynamo]: https://rockset.com/docs/amazon-dynamodb/
+
+## Secondary write paths
+There are some parts to the system that don't use the write path described
+above, because GitHub doesn't generate webhooks for them.
+
+Notably, this means that **installing the PyTorch Bot app to repos will not
+enable these write paths**. They must be manually wired up.
+
+#### Build artifacts
+Our build jobs create various artifacts that we want to save to persistent
+storage and use later. For example, most of our build jobs produce binary wheel
+distributions of PyTorch that are later downloaded by our test jobs.
+
+These are uploaded to S3 directly by the GitHub workflows in `pytorch/pytorch`.
+
+#### Logs and log classifications
+Logs are downloaded by the [`github-status-test`] AWS lambda, which receives
+webhooks through a manual connection with the `pytorch/pytorch` repo.
+`github-status-test` then pings another lambda, [`log-classifier`], which
+performs log classification (more detail in the [README]).
+
+[`github-status-test`]: https://us-east-1.console.aws.amazon.com/lambda/home?region=us-east-1#/functions/github-status-test?tab=code
```

Comment:
Is the lambda code versioned in test-infra or another repository? I'm trying to figure out where it is (excepts in the lambda itself) but haven't got any luck yet :)

Reply:
nope :) 

It would be a good better eng task to get it under source control